L62-core-call-credential-security-level.md

```diff
@@ -0,0 +1,121 @@
+languages, a call credential can only be communicated over a secure channel
+which is also acceptable from the security perspective. However with the
+introduction of local credential that can be used in both UDS and TCP loopback
+connections, any call credential (including the ones used to access Google APIs)
+can be transferred over a TCP loopback connection that is considered insecure, which
+greatly lowers the security bar. In more detals, UDS connections can be treated
+as relatively secure (assuming a correct directory permission is set on a file path)
+as a remote peer can be authenticated and data confidentiality can be guaranteed
+against passive adversaries. However for local TCP, we cannot authenticate a
+remote peer and thus both data confidentiality and integrity become meaningless.
+
+### Related Proposals:
+
+N/A
+
+## Proposal
+
+- A call credential should have an ability to dictate the security level of
+  a connection over which it will be communicated. It is an inherent property that
+  should be determined by its implementation, not the application that uses it.
+  If we give applications the power of setting security level, there will be risks that
+  applications could set it incorrectly either intentionally or by mistake.
+  Furthermore, by treating the security level as an inherent property the
+  behavior will be consistent with gRPC Java and Go implementations.
+
+  The security level should be one of privacy+integrity, integrity-only, and
+  insecure, which can be ordered from the highest (privacy+integrity) to lowest
+  (insecure). In case of a composite call credential, its security level should be set
+  to the highest level of any of its individual call credentials.
+
+- A security level should be also associated with each connection created by a channel
+  credential to indicate the level of transport security that connection is granted.
+  For a local connection, its security level will be set to privacy+integrity for UDS and
+  insecure for TCP loopback.
+
+
+- Before sending a call credential over a connection, we will extract the security levels
+  of call credential and the connection and check if the security level of
+  the connection is greater than or equal to that of call credential and only allow
+  the transfer of call credential if the above check succeeds. If the connection
+  does not meet the required security level, the call credential transfer will
+  fail. 
+
+
+### Add enum `grpc_security_level` to `include/grpc/grpc_security_constants.h`
+
+``` C
+
+/* Security level of grpc transport security */
+typedef enum {
+  GRPC_SECURITY_MIN = 0,
+  GRPC_SECURITY_NONE = GRPC_SECURITY_MIN,
+  GRPC_INTEGRITY_ONLY = 5,
+  GRPC_PRIVACY_AND_INTEGRITY = 10,
+  GRPC_SECURITY_MAX = 15,
+} grpc_security_level;
+
+```
+### Associate security level - `GRPC_PRIVACY_AND_INTEGRITY` to all call credentials used to access Google APIs.
+
+It will be realized by hard-coding `GRPC_PRIVACY_AND_INTEGRITY` in the
```

Comment:
Does this imply that privacy+integrity is going to be the default for all call creds, unless the implementation overrides it?  Will this affect backward compatibility for existing call creds that previously would work on any secure channel?

Reply:
Yes, that's the intention. 

Externally, we have SSL, TLS, and ALTS credentials and for all of connections created with those credentials, their security level will be privacy+integrity so there will not be backward compatibility issues. For local credentials, UDS will have privacy+integrity while Local TCP will have security_none levels but since it is currently under experimental namespace I think we reserve the right to make the changes.  Furthermore, we currently do not allow any call creds to be communicated over insecure channels, so we should be fine with insecure creds as well.  

Internally, I have already fixed all violations which used to send call credentials requiring privacy+integrity on integrity-only channels by elevating the channel's security level to privacy+integrity. So it should be fine as well.